File: cpp/面试题 17.05. 字母与数字.cpp
class Solution {
public:
    vector<string> findLongestSubarray(vector<string>& array) {
        int sum=0,n=array.size();
        unordered_map<int,int> mp;
        mp[0]=-1;
        int k=0;
        int maxL=0;
        for(int i=0;i<array.size();i++){
            sum+=isalpha(array[i][0])?1:-1;
            if(mp.count(sum)){
                int j=mp[sum];
                if(i-j>maxL){
                    maxL=i-j;
                    k=j+1;
                }
            }else{
               mp[sum]=i;
            }
        }
        return vector<string>(array.begin()+k,array.begin()+k+maxL);
    }
};

File: golang/面试题 17.05. 字母与数字.go
func findLongestSubarray(array []string) []string {
    sum:=0
    mp := map[int]int{}
    maxL := 0
    mp[0] = -1
    k := 0
    for i, s :=range array {
        if s[0] >= '0' && s[0] <= '9' {
            sum++
        } else {
            sum--
        }
        if j, ok := mp[sum]; ok {
            if i-j > maxL {
                maxL = i-j
                k = j+1
            }
        } else {
            mp[sum] = i
        }
    }
    return array[k : k+maxL]
}